### PR TITLE
HPCC-29721 setReplicateDir fails if config dir ends with path seperator

### PR DESCRIPTION
**Pull request by jakesmith**
Causing unittest CDaliDFSStressTests testDFSRename failures.

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: dali/base/dafdesc.cpp [hunk 1]
```diff
@@ -1721,7 +1721,10 @@ class CFileDescriptor:  public CFileDescriptorBase, implements ISuperFileDescrip
             if (!sc)
                 sc = getPathSepChar(dirname);
             StringBuffer tmp;
-            tmp.append(queryBaseDirectory(grp_unknown, 0, SepCharBaseOs(sc))).append(sc).append(s);
+            tmp.append(queryBaseDirectory(grp_unknown, 0, SepCharBaseOs(sc)));
+            if (sc != tmp.charAt(tmp.length()-1))
+                tmp.append(sc);
+            tmp.append(s);
```
Comment (jakesmith) [inline dali/base/dafdesc.cpp:1727]: NB: this is an incidental fix, observed while testing. A relative path and a config dir suffixed with a path separator caused a double path separator to be inserted.
I think this only ever happened (relative lfn dir path) via the unittests.

File: dali/base/dafdesc.cpp [hunk 2]
```diff
@@ -2934,30 +2937,33 @@ bool setReplicateDir(const char *dir,StringBuffer &out,bool isrep,const char *ba
...
+            out.append(r);
+            return true;
+        }
+        else if (isPathSepChar(dir[i])) // should mean that the last char of d was not a pathsepchar
+        {
             out.append(r).append(dir+i);
             return true;
         }
-    }
-    else if (count) { // this is a bit of a kludge to handle roxie backup
```
Comment (ghalliday) [inline dali/base/dafdesc.cpp:2948]: Did you mean to delete this code?  If so the count/match variables are no longer needed.

File: dali/base/dafdesc.cpp [hunk 3]
```diff
@@ -2934,30 +2937,33 @@ bool setReplicateDir(const char *dir,StringBuffer &out,bool isrep,const char *ba
...
             match = i;
             count++;
         }
+    }
     const char *r = repDir?repDir:queryBaseDirectory(grp_unknown, isrep ? 1 : 0,os);
-    if (d[i]==0) {
-        if ((dir[i]==0)||isPathSepChar(dir[i])) {
+    if (d[i]==0)
+    {
+        if (dir[i]==0)
```
Comment (ghalliday) [inline dali/base/dafdesc.cpp:2950]: Worth adding a comment to clarify the cases.  E.g.,
//directory is an exact match for the current base, so return replica unmodified.

File: dali/base/dafdesc.cpp [hunk 4]
```diff
@@ -2934,30 +2937,33 @@ bool setReplicateDir(const char *dir,StringBuffer &out,bool isrep,const char *ba
...
+        {
+            out.append(r);
+            return true;
+        }
+        else if (isPathSepChar(dir[i])) // should mean that the last char of d was not a pathsepchar
+        {
             out.append(r).append(dir+i);
```
Comment (ghalliday) [inline dali/base/dafdesc.cpp:2957]: does r definitely have no trailing /?  Otherwise this could lead to double //

**Review by ghalliday — COMMENTED** [review]
A few comments/questions.

File: dali/base/dafdesc.cpp (hunk 4 above)
Comment (jakesmith) [reply dali/base/dafdesc.cpp:2957]: that may be an additional issue. I'll check and probably add a check for tailing '/' on r too.

File: dali/base/dafdesc.cpp (hunk 2 above)
Comment (jakesmith) [reply dali/base/dafdesc.cpp:2948]: hm, I did not, well spotted.

**Comment by jakesmith** [comment]
@ghalliday - please review changes - I thought I pushed this last week..

**Review by ghalliday — APPROVED** [review]
Looks good, please squash

**Comment by jakesmith** [comment]
@ghalliday - squashed.

**Pull request merged**